pkg/volume/csi/csi_block.go

```diff
@@ -279,46 +200,121 @@ func (m *csiBlockMapper) MapDevice(devicePath, globalMapPath, volumeMapPath, vol
 	)
 
 	if err != nil {
-		glog.Errorf(log("blockMapper.MapDevice failed: %v", err))
-		if err := os.RemoveAll(dir); err != nil {
-			glog.Error(log("blockMapper.MapDevice failed to remove mapped dir after a NodePublish() error [%s]: %v", dir, err))
-		}
-		return err
+		glog.Errorf(log("blockMapper.publishVolumeForBlock failed: %v", err))
+		return "", err
 	}
 
-	return nil
+	return publishPath, nil
 }
 
-var _ volume.BlockVolumeUnmapper = &csiBlockMapper{}
-
-// TearDownDevice removes traces of the SetUpDevice.
-func (m *csiBlockMapper) TearDownDevice(globalMapPath, devicePath string) error {
+// SetUpDevice ensures the device is attached returns path where the device is located.
+func (m *csiBlockMapper) SetUpDevice() (string, error) {
 	if !m.plugin.blockEnabled {
-		return errors.New("CSIBlockVolume feature not enabled")
+		return "", errors.New("CSIBlockVolume feature not enabled")
 	}
+	glog.V(4).Infof(log("blockMapper.SetUpDevice called"))
 
-	glog.V(4).Infof(log("unmapper.TearDownDevice(globalMapPath=%s; devicePath=%s)", globalMapPath, devicePath))
+	// Get csiSource from spec
+	if m.spec == nil {
+		glog.Error(log("blockMapper.SetUpDevice spec is nil"))
+		return "", fmt.Errorf("spec is nil")
+	}
 
+	csiSource, err := getCSISourceFromSpec(m.spec)
+	if err != nil {
+		glog.Error(log("blockMapper.SetUpDevice failed to get CSI persistent source: %v", err))
+		return "", err
+	}
+
+	// Search for attachment by VolumeAttachment.Spec.Source.PersistentVolumeName
+	nodeName := string(m.plugin.host.GetNodeName())
+	attachID := getAttachmentName(csiSource.VolumeHandle, csiSource.Driver, nodeName)
+	attachment, err := m.k8s.StorageV1beta1().VolumeAttachments().Get(attachID, meta.GetOptions{})
+	if err != nil {
+		glog.Error(log("blockMapper.SetupDevice failed to get volume attachment [id=%v]: %v", attachID, err))
+		return "", err
+	}
+
+	if attachment == nil {
+		glog.Error(log("blockMapper.SetupDevice unable to find VolumeAttachment [id=%s]", attachID))
+		return "", errors.New("no existing VolumeAttachment found")
+	}
+
+	// Call NodeStageVolume
+	stagingPath, err := m.stageVolumeForBlock(csiSource, attachment)
+	if err != nil {
+		return "", err
+	}
+
+	// Call NodePublishVolume
+	publishPath, err := m.publishVolumeForBlock(csiSource, attachment, stagingPath)
+	if err != nil {
+		return "", err
```

Comment:
@rootfs

Thank you for your comment.

IIUC, reconciler will keep calling ```MountVolume``` to try to make ASW and DSW the same, even when ```MountVolume``` fails in ```SetUpDevice```. So, even if we call ```NodeUnstageVolume``` here to unstage the volume that failed in publish, another attempts to ```MountVolume``` would stage it again. (Then, ```MountVolume``` may fail in publish again.) In addition, if ```NodeUnstageVolume``` for Nth ```MountVolume``` are called before ```NodePublishVolume``` and after ```NodeStageVolume``` for N+1th ```MountVolume```, it will make ```NodePublishVolume``` call for unstaged volume, which violates CSI spec. Therefore, just calling ```NodeUnstageVolume``` here won't role back the status properly.

However, this raises me a question whether it is safe to keep the volume staged, after the failure in publish. At a glance, it seems safe from data corruption viewpoint, for we won't directly touch to the staged device and staging path.
The worst case senarios that I noticed are below:
  - If the pvc deletion is done before the publish-failed pod deletion, the device might be deleted while it is attached to a node,
  - If the pvc is used in another pod scheduled on a different node before the publish-failed pod deletion, the device might be attached to multiple nodes. 

@vladimirvivien

Is my understanding correct? And are there any issues that you can think of by not staging the volume there?